File: packages/sdk-go/storage.go
package pocketping

import (
	"context"
	"sync"
	"time"
)

// Storage is the interface for storage adapters.
// Implement this interface to use any database with PocketPing.
type Storage interface {
	// Session operations
	CreateSession(ctx context.Context, session *Session) error
	GetSession(ctx context.Context, sessionID string) (*Session, error)
	GetSessionByVisitorID(ctx context.Context, visitorID string) (*Session, error)
	UpdateSession(ctx context.Context, session *Session) error
	DeleteSession(ctx context.Context, sessionID string) error

	// Message operations
	SaveMessage(ctx context.Context, message *Message) error
	GetMessages(ctx context.Context, sessionID string, after string, limit int) ([]Message, error)
	GetMessage(ctx context.Context, messageID string) (*Message, error)

	// Optional cleanup
	CleanupOldSessions(ctx context.Context, olderThan time.Time) (int, error)
}

// StorageWithBridgeIDs extends Storage with bridge message ID operations.
// Implement this interface to support edit/delete synchronization with bridges.
type StorageWithBridgeIDs interface {
	Storage

	// UpdateMessage updates an existing message (for edit/delete).
	UpdateMessage(ctx context.Context, message *Message) error

	// SaveBridgeMessageIDs saves platform-specific message IDs for a message.
	SaveBridgeMessageIDs(ctx context.Context, messageID string, bridgeIDs BridgeMessageIds) error

	// GetBridgeMessageIDs retrieves platform-specific message IDs for a message.
	GetBridgeMessageIDs(ctx context.Context, messageID string) (*BridgeMessageIds, error)
}

// MemoryStorage is an in-memory storage adapter.
// Useful for development and testing. Data is lost on restart.
type MemoryStorage struct {
	mu              sync.RWMutex
	sessions        map[string]*Session
	messages        map[string][]Message  // sessionID -> messages
	messageByID     map[string]*Message   // messageID -> message
	bridgeMessageIDs map[string]*BridgeMessageIds // messageID -> bridge IDs
}

// NewMemoryStorage creates a new in-memory storage adapter.
func NewMemoryStorage() *MemoryStorage {
	return &MemoryStorage{
		sessions:         make(map[string]*Session),
		messages:         make(map[string][]Message),
		messageByID:      make(map[string]*Message),
		bridgeMessageIDs: make(map[string]*BridgeMessageIds),
	}
}

// CreateSession creates a new session.
func (m *MemoryStorage) CreateSession(ctx context.Context, session *Session) error {
	m.mu.Lock()
	defer m.mu.Unlock()

	m.sessions[session.ID] = session
	m.messages[session.ID] = []Message{}
	return nil
}

// GetSession retrieves a session by ID.
func (m *MemoryStorage) GetSession(ctx context.Context, sessionID string) (*Session, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	session, ok := m.sessions[sessionID]
	if !ok {
		return nil, nil
	}
	return session, nil
}

// GetSessionByVisitorID retrieves the most recent session for a visitor.
func (m *MemoryStorage) GetSessionByVisitorID(ctx context.Context, visitorID string) (*Session, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	var latest *Session
	for _, session := range m.sessions {
		if session.VisitorID == visitorID {
			if latest == nil || session.LastActivity.After(latest.LastActivity) {
				latest = session
			}
		}
	}
	return latest, nil
}

// UpdateSession updates an existing session.
func (m *MemoryStorage) UpdateSession(ctx context.Context, session *Session) error {
	m.mu.Lock()
	defer m.mu.Unlock()

	m.sessions[session.ID] = session
	return nil
}

// DeleteSession deletes a session and its messages.
func (m *MemoryStorage) DeleteSession(ctx context.Context, sessionID string) error {
	m.mu.Lock()
	defer m.mu.Unlock()

	// Remove messages for this session from messageByID
	if msgs, ok := m.messages[sessionID]; ok {
		for _, msg := range msgs {
			delete(m.messageByID, msg.ID)
		}
	}

	delete(m.sessions, sessionID)
	delete(m.messages, sessionID)
	return nil
}

// SaveMessage saves a message.
func (m *MemoryStorage) SaveMessage(ctx context.Context, message *Message) error {
	m.mu.Lock()
	defer m.mu.Unlock()

	// Check if message already exists (update case)
	if existing, ok := m.messageByID[message.ID]; ok {
		// Update existing message
		*existing = *message
		// Update in the slice too
		msgs := m.messages[message.SessionID]
		for i := range msgs {
			if msgs[i].ID == message.ID {
				msgs[i] = *message
				break
			}
		}
		return nil
	}

	// New message
	if _, ok := m.messages[message.SessionID]; !ok {
		m.messages[message.SessionID] = []Message{}
	}
	m.messages[message.SessionID] = append(m.messages[message.SessionID], *message)
	m.messageByID[message.ID] = message
	return nil
}

// GetMessages retrieves messages for a session.
func (m *MemoryStorage) GetMessages(ctx context.Context, sessionID string, after string, limit int) ([]Message, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	if limit <= 0 {
		limit = 50
	}

	msgs, ok := m.messages[sessionID]
	if !ok {
		return []Message{}, nil
	}

	startIndex := 0
	if after != "" {
		for i, msg := range msgs {
			if msg.ID == after {
				startIndex = i + 1
				break
			}
		}
	}

	endIndex := startIndex + limit
	if endIndex > len(msgs) {
		endIndex = len(msgs)
	}

	result := make([]Message, endIndex-startIndex)
	copy(result, msgs[startIndex:endIndex])
	return result, nil
}

// GetMessage retrieves a message by ID.
func (m *MemoryStorage) GetMessage(ctx context.Context, messageID string) (*Message, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	msg, ok := m.messageByID[messageID]
	if !ok {
		return nil, nil
	}
	return msg, nil
}

// CleanupOldSessions removes sessions older than the given time.
func (m *MemoryStorage) CleanupOldSessions(ctx context.Context, olderThan time.Time) (int, error) {
	m.mu.Lock()
	defer m.mu.Unlock()

	count := 0
	toDelete := []string{}

	for id, session := range m.sessions {
		if session.LastActivity.Before(olderThan) {
			toDelete = append(toDelete, id)
			count++
		}
	}

	for _, id := range toDelete {
		// Remove messages for this session from messageByID
		if msgs, ok := m.messages[id]; ok {
			for _, msg := range msgs {
				delete(m.messageByID, msg.ID)
			}
		}
		delete(m.sessions, id)
		delete(m.messages, id)
	}

	return count, nil
}

// GetAllSessions returns all sessions. Useful for admin/debug.
func (m *MemoryStorage) GetAllSessions(ctx context.Context) ([]*Session, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	sessions := make([]*Session, 0, len(m.sessions))
	for _, session := range m.sessions {
		sessions = append(sessions, session)
	}
	return sessions, nil
}

// GetSessionCount returns the total number of sessions.
func (m *MemoryStorage) GetSessionCount(ctx context.Context) (int, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	return len(m.sessions), nil
}

// UpdateMessage updates an existing message (for edit/delete).
func (m *MemoryStorage) UpdateMessage(ctx context.Context, message *Message) error {
	m.mu.Lock()
	defer m.mu.Unlock()

	if _, ok := m.messageByID[message.ID]; !ok {
		return nil // Message doesn't exist
	}

	// Update in messageByID
	m.messageByID[message.ID] = message

	// Update in the slice too
	msgs := m.messages[message.SessionID]
	for i := range msgs {
		if msgs[i].ID == message.ID {
			msgs[i] = *message
			break
		}
	}

	return nil
}

// SaveBridgeMessageIDs saves platform-specific message IDs for a message.
func (m *MemoryStorage) SaveBridgeMessageIDs(ctx context.Context, messageID string, bridgeIDs BridgeMessageIds) error {
	m.mu.Lock()
	defer m.mu.Unlock()

	existing := m.bridgeMessageIDs[messageID]
	if existing != nil {
		// Merge with existing
		if bridgeIDs.TelegramMessageID != 0 {
			existing.TelegramMessageID = bridgeIDs.TelegramMessageID
		}
		if bridgeIDs.DiscordMessageID != "" {
			existing.DiscordMessageID = bridgeIDs.DiscordMessageID
		}
		if bridgeIDs.SlackMessageTS != "" {
			existing.SlackMessageTS = bridgeIDs.SlackMessageTS
		}
	} else {
		m.bridgeMessageIDs[messageID] = &bridgeIDs
	}

	return nil
}

// GetBridgeMessageIDs retrieves platform-specific message IDs for a message.
func (m *MemoryStorage) GetBridgeMessageIDs(ctx context.Context, messageID string) (*BridgeMessageIds, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	return m.bridgeMessageIDs[messageID], nil
}

// Ensure MemoryStorage implements Storage interface
var _ Storage = (*MemoryStorage)(nil)

// Ensure MemoryStorage implements StorageWithBridgeIDs interface
var _ StorageWithBridgeIDs = (*MemoryStorage)(nil)

File: packages/sdk-ruby/lib/pocketping/storage.rb
# frozen_string_literal: true

require "time"

module PocketPing
  module Storage
    # Abstract base class for storage adapters
    #
    # Implement this interface to create custom storage backends
    # (PostgreSQL, Redis, MongoDB, etc.)
    #
    # @abstract Subclass and override the abstract methods
    #
    # @example Creating a custom storage adapter
    #   class RedisStorage < PocketPing::Storage::Base
    #     def initialize(redis_client)
    #       @redis = redis_client
    #     end
    #
    #     def create_session(session)
    #       @redis.set("session:#{session.id}", session.to_json)
    #     end
    #
    #     # ... implement other methods
    #   end
    class Base
      # Create a new session
      #
      # @param session [Session] The session to create
      # @return [void]
      def create_session(session)
        raise NotImplementedError, "#{self.class} must implement #create_session"
      end

      # Get a session by ID
      #
      # @param session_id [String] The session ID
      # @return [Session, nil] The session or nil if not found
      def get_session(session_id)
        raise NotImplementedError, "#{self.class} must implement #get_session"
      end

      # Update an existing session
      #
      # @param session [Session] The session to update
      # @return [void]
      def update_session(session)
        raise NotImplementedError, "#{self.class} must implement #update_session"
      end

      # Delete a session
      #
      # @param session_id [String] The session ID
      # @return [void]
      def delete_session(session_id)
        raise NotImplementedError, "#{self.class} must implement #delete_session"
      end

      # Save a message
      #
      # @param message [Message] The message to save
      # @return [void]
      def save_message(message)
        raise NotImplementedError, "#{self.class} must implement #save_message"
      end

      # Get messages for a session
      #
      # @param session_id [String] The session ID
      # @param after [String, nil] Return messages after this message ID
      # @param limit [Integer] Maximum number of messages to return
      # @return [Array<Message>] List of messages
      def get_messages(session_id, after: nil, limit: 50)
        raise NotImplementedError, "#{self.class} must implement #get_messages"
      end

      # Get a message by ID
      #
      # @param message_id [String] The message ID
      # @return [Message, nil] The message or nil if not found
      def get_message(message_id)
        raise NotImplementedError, "#{self.class} must implement #get_message"
      end

      # Clean up old sessions
      #
      # @param older_than [Time] Delete sessions with last_activity before this time
      # @return [Integer] Number of sessions deleted
      def cleanup_old_sessions(older_than)
        0
      end

      # Get the most recent session for a visitor
      #
      # @param visitor_id [String] The visitor ID
      # @return [Session, nil] The session or nil if not found
      def get_session_by_visitor_id(visitor_id)
        nil
      end

      # Update an existing message (for edit/delete)
      # Optional: defaults to save_message
      #
      # @param message [Message] The message to update
      # @return [void]
      def update_message(message)
        save_message(message)
      end

      # Save platform-specific message IDs for a message
      # Optional: implement for edit/delete synchronization with bridges
      #
      # @param message_id [String] The message ID
      # @param bridge_ids [BridgeMessageIds] The bridge message IDs
      # @return [void]
      def save_bridge_message_ids(message_id, bridge_ids)
        nil
      end

      # Get platform-specific message IDs for a message
      # Optional: implement for edit/delete synchronization with bridges
      #
      # @param message_id [String] The message ID
      # @return [BridgeMessageIds, nil] The bridge message IDs or nil
      def get_bridge_message_ids(message_id)
        nil
      end
    end

    # In-memory storage adapter
    #
    # Useful for development, testing, and simple deployments.
    # Note: Data is lost when the process restarts.
    #
    # @example
    #   storage = PocketPing::Storage::MemoryStorage.new
    #   pp = PocketPing::Client.new(storage: storage)
    class MemoryStorage < Base
      def initialize
        @sessions = {}
        @messages = {}
        @message_by_id = {}
        @bridge_message_ids = {}
        @mutex = Mutex.new
      end

      # @see Base#create_session
      def create_session(session)
        @mutex.synchronize do
          @sessions[session.id] = session
          @messages[session.id] = []
        end
        nil
      end

      # @see Base#get_session
      def get_session(session_id)
        @mutex.synchronize do
          @sessions[session_id]
        end
      end

      # @see Base#update_session
      def update_session(session)
        @mutex.synchronize do
          @sessions[session.id] = session
        end
        nil
      end

      # @see Base#delete_session
      def delete_session(session_id)
        @mutex.synchronize do
          @sessions.delete(session_id)
          messages = @messages.delete(session_id) || []
          messages.each { |msg| @message_by_id.delete(msg.id) }
        end
        nil
      end

      # @see Base#save_message
      def save_message(message)
        @mutex.synchronize do
          @messages[message.session_id] ||= []

          # Check if message already exists (update case)
          existing_index = @messages[message.session_id].index { |m| m.id == message.id }
          if existing_index
            @messages[message.session_id][existing_index] = message
          else
            @messages[message.session_id] << message
          end

          @message_by_id[message.id] = message
        end
        nil
      end

      # @see Base#get_messages
      def get_messages(session_id, after: nil, limit: 50)
        @mutex.synchronize do
          messages = @messages[session_id] || []

          if after
            start_index = 0
            messages.each_with_index do |msg, i|
              if msg.id == after
                start_index = i + 1
                break
              end
            end
            messages = messages[start_index..] || []
          end

          messages.first(limit)
        end
      end

      # @see Base#get_message
      def get_message(message_id)
        @mutex.synchronize do
          @message_by_id[message_id]
        end
      end

      # @see Base#cleanup_old_sessions
      def cleanup_old_sessions(older_than)
        count = 0
        to_delete = []

        @mutex.synchronize do
          @sessions.each do |session_id, session|
            if session.last_activity < older_than
              to_delete << session_id
            end
          end
        end

        to_delete.each do |session_id|
          delete_session(session_id)
          count += 1
        end

        count
      end

      # @see Base#get_session_by_visitor_id
      def get_session_by_visitor_id(visitor_id)
        @mutex.synchronize do
          visitor_sessions = @sessions.values.select { |s| s.visitor_id == visitor_id }
          return nil if visitor_sessions.empty?

          # Return most recent by last_activity
          visitor_sessions.max_by(&:last_activity)
        end
      end

      # Get all sessions (useful for admin/debug)
      #
      # @return [Array<Session>] All sessions
      def get_all_sessions
        @mutex.synchronize do
          @sessions.values.dup
        end
      end

      # Get total session count
      #
      # @return [Integer] Number of sessions
      def get_session_count
        @mutex.synchronize do
          @sessions.size
        end
      end

      # Clear all data (useful for testing)
      #
      # @return [void]
      def clear!
        @mutex.synchronize do
          @sessions.clear
          @messages.clear
          @message_by_id.clear
          @bridge_message_ids.clear
        end
        nil
      end

      # @see Base#update_message
      def update_message(message)
        @mutex.synchronize do
          return unless @message_by_id.key?(message.id)

          # Update in message_by_id
          @message_by_id[message.id] = message

          # Update in the session's messages array
          if @messages[message.session_id]
            existing_index = @messages[message.session_id].index { |m| m.id == message.id }
            @messages[message.session_id][existing_index] = message if existing_index
          end
        end
        nil
      end

      # @see Base#save_bridge_message_ids
      def save_bridge_message_ids(message_id, bridge_ids)
        @mutex.synchronize do
          existing = @bridge_message_ids[message_id]
          if existing
            @bridge_message_ids[message_id] = existing.merge_with(bridge_ids)
          else
            @bridge_message_ids[message_id] = bridge_ids
          end
        end
        nil
      end

      # @see Base#get_bridge_message_ids
      def get_bridge_message_ids(message_id)
        @mutex.synchronize do
          @bridge_message_ids[message_id]
        end
      end
    end
  end
end
